src/widget/wstemcontrol.cpp

```diff
@@ -0,0 +1,166 @@
+#include "widget/wstemcontrol.h"
+
+#include <QCoreApplication>
+#include <QDragEnterEvent>
+#include <QEvent>
+#include <QHBoxLayout>
+#include <QLabel>
+#include <QMainWindow>
+#include <QPalette>
+#include <QPushButton>
+#include <QStyleOption>
+#include <QVBoxLayout>
+
+#include "control/controlproxy.h"
+#include "moc_wstemcontrol.cpp"
+#include "track/track.h"
+#include "util/math.h"
+#include "util/parented_ptr.h"
+#include "widget/controlwidgetconnection.h"
+#include "widget/wglwidget.h"
+#include "widget/wknob.h"
+#include "widget/wknobcomposed.h"
+#include "widget/wlabel.h"
+
+namespace {
+// FIXME(XXX) this is a workaround to ensure that that knob keep the aspect
+// ratio as defined in its maxSize. This needs to be moved an implemented by the
+// component directly
+void ensureWidgetRatio(QLayout* layout) {
+    for (int i = 0; i < layout->count(); i++) {
+        auto w = layout->itemAt(i)->widget();
+        if (!qobject_cast<WKnobComposed*>(w) && !qobject_cast<WKnob*>(w)) {
+            continue;
+        }
+        float ratio = static_cast<float>(w->maximumWidth()) /
+                static_cast<float>(w->maximumHeight());
+        int maxSize = qMin(w->width(), w->height());
```

Comment:
I do not understand this. Can you add a source code comment?

Reply:
This is a nasty hack to try and keep the knob ratio constant to 1, in this case `maxSize` takes the value of the lower dimension.
I'm planing to remove `wstemcontrol.cpp` once this PR is ready as there is a few issue with the UI, so this can be moved in a subsequent PR/dropped entirely.